Add Server Side YAML Restrictions

jjulien opened the pull request:
Added new --allow-restricted-repo-config and --repo-config options

This allows usage of a server side repo config file to restrict certain fields from being used in a repos atlantis.yaml file.

When this feature is activated apply_requirements, workflow, and workflows can only be specified in an atlantis.yaml file if explicitly allowed by the server side repo config.

The repo config file provides the ability to specify a default set of workflows, and default values for apply_requirements and workflow to use use on a per repo basis.  It also supports applying to a collection of repos by using regex to match a repo name.

If more than one repo name matches, the values from last repo matched are used.

Closes #47

lkysow commented:
Hi John, thanks for the amazing PR! I'm hoping to get some time this week to review.

kipkoan commented:
This is an implementation of the [[RFC] Server-side atlantis.yaml
](https://docs.google.com/document/d/1w2wePfXGA3L151Af6D0kB1aKVB7pXys_l9Vk9SD7kyA/edit).

@jjulien - can you comment as to what may be different or missing in this implementation from what is documented in the RFC?

jjulien commented:
@kipkoan I followed the RFC with the exception of two changes that I commented on in the RFC doc but were never officially included.

1. Moved the new repo config options into their own file rather than including them in the server `config.yaml`.  The new repo config is specified using the CLI option `--repo-config`.  The file implements the `repos` and `workflows` keys using the schema defined in the RFC.
1. Deprecated rather than removed `--allow-repo-config`, and implemented the repo restrictions as new functionality.  The new functionality is triggered when the `--allow-restricted-repo-config` flag is used.  The `--allow-repo-config` flag is now marked as deprecated (which prevents it from showing up in help as implemented by Cobra), and only one of either `--allow-restricted-repo-config` or `--allow-repo-config` is allowed to be used.

TraGicCode commented:
Hey @jjulien @ikysow

While I haven’t looked at the code I read the RFC and really excited about this feature.  This will address our concerns I’ve been having in regards to any pull request creator automatically having sudo “so to speak” on the Atlantis server which makes it hard to justify with the security team as “we just have to trust Users at our org will be not be malicious” in an enterprise environment but theoretically could take down production and/or expose sensitive secrets from terraforms remote state storage.

jjulien commented:
@lkysow This is ready to go now with the changes you mentioned in the RFC.

- Removal of `--allow-restricted-repo-config` option, and making this the default behavior.
- Deprecated `--allow-repo-config`

lkysow commented:
Thanks @jjulien! I'm going to pull this into my own branch and make some finishing touches.
Can you squash your commits into 1 please.

jjulien commented:
@lkysow Commits have been squashed.  Thanks!

lkysow commented:
Thanks! I'm going to merge something that might cause conflicts here but don't worry about fixing it, I will handle that.

lkysow commented:
Hi all subscribers,
If you're interested in trying out server-side config, I've created an alpha release off of @jjulien's original work and my refinement. You can get it here: https://github.com/runatlantis/atlantis/releases/tag/v0.7.0-alpha1 or from Docker: runatlantis/atlantis:v0.7.0-alpha.

• You can read docs about it here: https://deploy-preview-546--runatlantis.netlify.com/docs/repos-yaml-reference.html#overview
• If you've got feedback, either reply here or open up an issue. Thanks!

lkysow commented:
@jjulien I've refactored your changes and pushed everything to a release-0.7 branch because I don't want to merge into master and have the website update. If you're curious you can see the diff here: https://github.com/runatlantis/atlantis/compare/release-0.7?expand=1

Once 0.7 is ready I'll merge that branch into master.

I'm going to close this due to ☝️. Thanks for all your hard work 🙏! Should hopefully get a release out soon.
